lib/ansible/module_utils/oneview.py

```diff
@@ -0,0 +1,782 @@
+        """
+
+        argument_spec = self.__build_argument_spec(additional_arg_spec, validate_etag_support)
+
+        self.module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False)
+
+        self.__check_hpe_oneview_sdk()
+        self.__create_oneview_client()
+
+        self.state = self.module.params.get('state')
+        self.data = self.module.params.get('data')
+
+        # Preload params for get_all - used by facts
+        self.facts_params = self.module.params.get('params') or {}
+
+        # Preload options as dict - used by facts
+        self.options = self.transform_list_to_dict(self.module.params.get('options'))
+
+        self.validate_etag_support = validate_etag_support
+
+    def __build_argument_spec(self, additional_arg_spec, validate_etag_support):
+
+        merged_arg_spec = dict()
+        merged_arg_spec.update(self.ONEVIEW_COMMON_ARGS)
+
+        if validate_etag_support:
+            merged_arg_spec.update(self.ONEVIEW_VALIDATE_ETAG_ARGS)
+
+        if additional_arg_spec:
+            merged_arg_spec.update(additional_arg_spec)
+
+        return merged_arg_spec
+
+    def __check_hpe_oneview_sdk(self):
+        if not HAS_HPE_ONEVIEW:
+            self.module.fail_json(msg=self.HPE_ONEVIEW_SDK_REQUIRED)
+        if not HAS_FUTURE:
+            self.module.fail_json(msg=self.FUTURE_PACKAGE_REQUIRED)
+
+    def __create_oneview_client(self):
+        if not self.module.params['config']:
+            self.oneview_client = OneViewClient.from_environment_variables()
+        else:
+            self.oneview_client = OneViewClient.from_json_file(self.module.params['config'])
+
+    def execute_module(self):
+        """
+        Abstract function, must be implemented by the inheritor.
+
+        This method is called from the run method. It should contains the module logic
+
+        Returns:
+            dict:
+                 It must return a dictionary with the attributes for the module result,
+                 such as ansible_facts, msg and changed.
+        """
+        raise HPOneViewException("execute_module not implemented")
+
+    def run(self):
+        """
+        Common implementation of the OneView run modules.
+        It calls the inheritor 'execute_module' function and sends the return to the Ansible.
+        It handles any HPOneViewException in order to signal a failure to Ansible, with a descriptive error message.
+        """
+        try:
+            if self.validate_etag_support:
+                if not self.module.params.get('validate_etag'):
+                    self.oneview_client.connection.disable_etag_validation()
+
+            result = self.execute_module()
+
+            if "changed" not in result:
+                result['changed'] = False
+
+            self.module.exit_json(**result)
+
+        except HPOneViewException as exception:
+            self.module.fail_json(msg='; '.join(str(e) for e in exception.args))
```

Comment:
Also note, str can traceback if e contains non-ascii.  Better to use the to_native function from ansible.module_utils._text.